lib/os_io.c

```diff
@@ -8901,6 +8901,263 @@ ___SCMOBJ options;)
+___BOOL for_writing;
+int i;
+int pass;
+___device_select_state *state;)
+{
+  ___device_raw *d = ___CAST(___device_raw*,self);
+
+#ifdef USE_POSIX
+  int stage = (for_writing
+               ? d->base.write_stage
+               : d->base.read_stage);
+
+  if (pass == ___SELECT_PASS_1)
+    {
+      if (stage != ___STAGE_OPEN)
+        state->timeout = ___time_mod.time_neg_infinity;
+      else
+        ___device_select_add_fd (state, d->fd, for_writing);
+
+      return ___FIX(___SELECT_SETUP_DONE);
+    }
+
+  /* pass == ___SELECT_PASS_CHECK */
+  if (stage != ___STAGE_OPEN)
+    state->devs[i] = NULL;
+  else if (for_writing
+           ? ___FD_ISSET(d->fd, &state->writefds)
+           : ___FD_ISSET(d->fd, &state->readfds))
+    state->devs[i] = NULL;
+  
+  return ___FIX(___NO_ERR);
+#endif
+
+  return ___FIX(___UNIMPL_ERR);
+}
+
+___HIDDEN ___SCMOBJ ___device_raw_release_virt
+   ___P((___device *self),
+        (self)
+___device *self;)
+{
+  return ___FIX(___NO_ERR);
+}
+
+___HIDDEN ___SCMOBJ ___device_raw_force_output_virt
+   ___P((___device *self,
+         int level),
+        (self,
+         level)
+___device *self;
+int level;)
+{
+  return ___FIX(___NO_ERR);
+}
+
+___HIDDEN ___device_raw_vtbl ___device_raw_table =
+{
+  {
+    ___device_raw_kind,
+    ___device_raw_select_virt,
+    ___device_raw_release_virt,
+    ___device_raw_force_output_virt,
+    ___device_raw_close_virt
+  }
+};
+
+#ifdef USE_POSIX
+
+___SCMOBJ ___device_raw_setup_from_fd
+   ___P((___device_raw **dev,
+         ___device_group *dgroup,
+         int fd,
+         int direction),
+        (dev,
+         dgroup,
+         fd,
+         direction)
+___device_raw **dev;
+___device_group *dgroup;
+int fd;
+int direction;)
+{
+  ___device_raw *d;
+
+  d = ___CAST(___device_raw*,
+              ___ALLOC_MEM(sizeof (___device_raw)));
+
+  if (d == NULL)
+    return ___FIX(___HEAP_OVERFLOW_ERR);
+
+
+  d->base.vtbl = &___device_raw_table;
+  d->base.refcount = 1;
+  d->base.direction = direction;
+  d->base.close_direction = 0; /* prevent closing on errors */
+
+  if (direction & ___DIRECTION_RD)
+    d->base.read_stage = ___STAGE_OPEN;
+  else
+    d->base.read_stage = ___STAGE_CLOSED;
+
+  if (direction & ___DIRECTION_WR)
+    d->base.write_stage = ___STAGE_OPEN;
+  else
+    d->base.write_stage = ___STAGE_CLOSED;
+
+  d->fd = fd;
+
+  device_transfer_close_responsibility (___CAST(___device*,d));
+
+  *dev = d;
+
+  ___device_add_to_group (dgroup, &d->base);
+
+  return ___FIX(___NO_ERR);
+}
+#endif
+
+___SCMOBJ ___os_device_raw_open
+   ___P((___SCMOBJ index,
+         ___SCMOBJ flags),
+        (index,
+         flags)
+___SCMOBJ fd;
+___SCMOBJ flags;)
+{
+
+#ifdef USE_POSIX
+  ___SCMOBJ e;
+  ___device_raw *dev;
+  ___SCMOBJ result;
+
+  int fd;
+  int fl;
+  int direction;
+
+  device_translate_flags (___INT(flags),
+                          &fl,
+                          &direction);
+
+  fd = ___INT(index);
+
+  if ((e = ___device_raw_setup_from_fd
```

Comment:
Indent properly... open paren for parameters should be 2 spaces indented from the ```___device_...```.